apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore/Trace/CloudTraceExtension.cs

```diff
@@ -0,0 +1,174 @@
+            app.UseMiddleware<CloudTraceMiddleware>();
+        }
+
+        /// <summary>
+        /// Adds the needed services for Google Cloud Tracing. Used with <see cref="UseGoogleTrace"/>.
+        /// </summary>
+        /// <param name="projectId">The Google Cloud Platform project ID. Cannot be null.</param>
+        /// <param name="config">Optional trace configuration, if unset the default will be used.</param>
+        /// <param name="clientTask">Optional task which produces the Trace client, if 
+        ///     unset the default will be used.</param>
+        public static void AddGoogleTrace(
+            this IServiceCollection services, string projectId,
+            TraceConfiguration config = null, Task<TraceServiceClient> clientTask = null)
+        {
+            GaxPreconditions.CheckNotNull(services, nameof(services));
+            GaxPreconditions.CheckNotNull(projectId, nameof(projectId));
+
+            clientTask = clientTask ?? TraceServiceClient.CreateAsync();
+            config = config ?? TraceConfiguration.Create();
+
+            IConsumer<TraceProto> consumer = ConsumerFactory<TraceProto>.GetConsumer(
+                 new GrpcTraceConsumer(clientTask), TraceSizer.Instance, config.BufferOptions);
+
+            services.AddSingleton(new ProjectId { Id = projectId });
+            services.AddSingleton(consumer);
+            services.AddSingleton(TraceIdFactory.Create());
+            services.AddSingleton(RateLimitingTraceOptionsFactory.Create(config));
+            services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
+
+            services.AddScoped(CreateTraceHeaderContext);
+            services.AddScoped(CreateManagedTracer);
+        }
+
+        /// <summary>
+        /// Creates an <see cref="TraceHeaderContext"/> based on the current <see cref="HttpContext"/>.
+        /// </summary>
+        internal static TraceHeaderContext CreateTraceHeaderContext(IServiceProvider provider)
+        {
+            var accessor = provider.GetService<IHttpContextAccessor>();
+            string header = accessor?.HttpContext?.Request?.Headers[TraceHeaderContext.TraceHeader];
+            return TraceHeaderContext.FromHeader(header);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="IManagedTracer"/> based on the <see cref="TraceHeaderContext"/> and
+        /// the rate limiter.
+        /// </summary>
+        internal static IManagedTracer CreateManagedTracer(IServiceProvider provider)
+        {
+            var headerContext = provider.GetService<TraceHeaderContext>();
+            var rateLimitingFactory = provider.GetService<RateLimitingTraceOptionsFactory>();
+            var projectId = provider.GetService<ProjectId>();
+            var traceIdFactory = provider.GetService<TraceIdFactory>();
+            var consumer = provider.GetService<IConsumer<TraceProto>>();
+
+            // Check that we have all the needed services.
```

Comment:
Possibly 5 calls to GaxPreconditions.CheckState? Would make it easier to work out what's wrong.